### PR TITLE
[4.0] Diagnostics for Minimal and Complete Patterns

### PR DESCRIPTION
**Pull request by CodaFi**
Redoes a phase of type checking to validate patterns for minimality and completeness.

Cherry-picked from #8908, #9121, #9155, and #9180.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
Wow, this cherry-pick was pretty bad.  I'm gonna rebase and try again.

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -3566,6 +3558,23 @@ WARNING(debug_long_closure_body, none,
...
+//------------------------------------------------------------------------------
+
+
+ERROR(empty_switch_stmt,none,
+      "'switch' statement body must have at least one 'case' or 'default' "
+      "block; do you want to add a default case?",())
+ERROR(non_exhaustive_switch,none,
+      "%select{switch must be exhaustive, consider adding |do you want to add }0"
+      "%select{missing cases|a default clause}1"
+      "%select{:|?}0 %2", (bool, bool, StringRef))
```
Comment (nkcsgexi) [inline include/swift/AST/DiagnosticsSema.def]: Could you drop the last StringRef? we don't need to duplicate the fixit content.

**Review by nkcsgexi — CHANGES_REQUESTED** [review]

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Ah yeah.  That was leftover from when I thought I could combine these diagnostics.  

Thanks for catching this, I'll shoot a quick pull request over and cherry pick that into here.

Comment (nkcsgexi) [inline include/swift/AST/DiagnosticsSema.def]: Fixing on master first will be awesome. Thank you!

Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Ah yeah.  That was leftover from when I thought I could combine this fixit with the other one. 

Thanks!  I'll shoot a quick pull request over and cherry-pick that in here.

**Comment by CodaFi** [comment]
Weird: If you force-push at a branch that becomes even with the one you're targeting a pull request at, it closes it.

**Comment by CodaFi** [comment]
@nkcsgexi These commits are going to be auto-merged anyways, right?  I don't think this PR is necessary.

**Comment by nkcsgexi** [comment]
@CodaFi it seems they won't be auto-merged to swift-4.0-branch, shall we revive the pull request?

**Pull request closed without merging**